server/src/main/java/org/elasticsearch/index/store/FsDirectoryService.java

```diff
@@ -97,11 +92,31 @@ protected Directory newFSDirectory(Path location, LockFactory lockFactory) throw
                 // Use Lucene defaults
                 final FSDirectory primaryDirectory = FSDirectory.open(location, lockFactory);
                 if (primaryDirectory instanceof MMapDirectory) {
-                    return new FileSwitchDirectory(PRIMARY_EXTENSIONS, primaryDirectory, new NIOFSDirectory(location, lockFactory), true) {
+                    return new NIOFSDirectory(location, lockFactory) {
+                        @Override
+                        public IndexInput openInput(String name, IOContext context) throws IOException {
+                            String extension = FileSwitchDirectory.getExtension(name);
+                            switch(extension) {
+                                // We are mmapping norms, docvalues as well as term dictionaries, all other files are served through NIOFS
+                                // this provides good random access performance and does not lead to page cache thrashing.
+                                case "nvd":
+                                case "dvd":
+                                case "tim":
+                                    ensureOpen();
+                                    ensureCanRead(name);
```

Comment:
As far as I could see these checks are performed by `MMapDirectory#openInput()` as well so they're done twice now?

Reply:
no we need to do these check on the outer directory since the inner doesn't know about hte pending deletes.